number_theory/modular.h

```diff
@@ -111,6 +112,120 @@ class Modular {
+  bool operator OP(const M &lhs, const T &rhs) {                  \
+    return lhs.METHOD(static_cast<M>(rhs));                       \
+  }                                                               \
+  template <CONCEPT M, std::convertible_to<M> T>                  \
+  bool operator OP(const T &lhs, const M &rhs) {                  \
+    return static_cast<M>(lhs).METHOD(rhs);                       \
+  }                                                               \
+  template <CONCEPT M>                                            \
+  bool operator OP_NEG(const M &lhs, const M &rhs) {              \
+    return !lhs.METHOD(rhs);                                      \
+  }                                                               \
+  template <CONCEPT M, std::convertible_to<M> T>                  \
+  bool operator OP_NEG(const M &lhs, const T &rhs) {              \
+    return !lhs.METHOD(static_cast<M>(rhs));                      \
+  }                                                               \
+  template <CONCEPT M, std::convertible_to<M> T>                  \
+  bool operator OP_NEG(const T &lhs, const M &rhs) {              \
+    return !static_cast<M>(lhs).METHOD(rhs);                      \
+  }
+
+// Overloads an unary operator with a class method
+#define OVERLOAD_UNARY_OPERRATOR(CONCEPT, OP, METHOD) \
+  template <CONCEPT M>                                \
+  M operator OP(const M &x) {                         \
+    return x.METHOD();                                \
+  }
+
+// Overload an increment/decrement operator with a class method
+#define OVERLOAD_INCDEC_OPERRATOR(CONCEPT, OP, METHOD) \
+  /* prefix */                                         \
+  template <CONCEPT M>                                 \
+  M &operator OP(M &x) {                               \
+    x = x.METHOD(M(1));                                \
+    return x;                                          \
+  }                                                    \
+  /* postfix */                                        \
+  template <CONCEPT M>                                 \
+  M operator OP(M &x, int) {                           \
+    M old = x;                                         \
+    x = x.METHOD(M(1));                                \
+    return old;                                        \
+  }
+
+OVERLOAD_ARITHMETIC_OPERATOR(IsModular, +, add)
+OVERLOAD_ARITHMETIC_OPERATOR(IsModular, -, subtract)
+OVERLOAD_ARITHMETIC_OPERATOR(IsModular, *, multiply)
+
+OVERLOAD_INPLACE_OPERATOR(IsModular, +=, add)
+OVERLOAD_INPLACE_OPERATOR(IsModular, -=, subtract)
+OVERLOAD_INPLACE_OPERATOR(IsModular, *=, multiply)
+
+OVERLOAD_COMPARISON_OPERATOR(IsModular, ==, !=, equal)
+
+OVERLOAD_UNARY_OPERRATOR(IsModular, -, negate)
+OVERLOAD_UNARY_OPERRATOR(IsModular, +, get)
+
+OVERLOAD_INCDEC_OPERRATOR(IsModular, ++, add)
+OVERLOAD_INCDEC_OPERRATOR(IsModular, --, subtract)
+
+#undef OVERLOAD_ARITHMETIC_OPERATOR
```

Comment:
What's the advantage of undefining them other than making them unavailable for future calls?

Reply:
We undefine them to avoid naming conflicts with other libraries.